File: leetcode/231.rs
/*
Question:
Given an integer n, return true if it is a power of two. Otherwise, return false.
An integer n is a power of two, if there exists an integer x such that n == 2x.
*/

// Approach:
// If the binary form of the number only has a single 1 bit then it is a power of two
// 1) We first check if the number is positive as a negetive number is not a power of two
// 2) Then we convert the number to a binary string
// 3) Then we filter the string to keep only 1 bits
// 4) Then we check if there is exactly 1 positive bit and if there is then we return true, otherwise false

impl Solution {
    pub fn is_power_of_two(n: i32) -> bool {
        if n > 0 && format!("{:b}",n).chars().filter(|x| *x == '1').count() == 1 {true} else {false}
    }
}

File: leetcode/231.java
/*
Question:
Given an integer n, return true if it is a power of two. Otherwise, return false.

An integer n is a power of two, if there exists an integer x such that n == 2x.
*/

// Approach:
// 1) If n is less than or equal to 0, return false
// 2) If n is power of 2, then n & (n-1) will be 0
//     So, return (n & (n-1)) == 0

class Solution {
  public boolean isPowerOfTwo(int n) {
    if(n <= 0) return false;
    return (n & (n-1)) == 0;
  }
}
